sklearn/naive_bayes.py

```diff
@@ -958,3 +985,245 @@ def _joint_log_likelihood(self, X):
+
+        This is especially useful when the whole dataset is too big to fit in
+        memory at once.
+
+        This method has some performance overhead hence it is better to call
+        partial_fit on chunks of data that are as large as possible
+        (as long as fitting in the memory budget) to hide the overhead.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples and
+            n_features is the number of features. Here, each feature of X is
+            assumed to be from a different categorical distribution.
+
+        y : array-like, shape = [n_samples]
+            Target values.
+
+        classes : array-like, shape = [n_classes] (default=None)
+            List of all the classes that can possibly appear in the y vector.
+
+            Must be provided at the first call to partial_fit, can be omitted
+            in subsequent calls.
+
+        sample_weight : array-like, shape = [n_samples], (default=None)
+            Weights applied to individual samples (1. for unweighted).
+
+        Returns
+        -------
+        self : object
+        """
+        self._check_settings()
+        return (super(CategoricalNB, self)
+                .partial_fit(X, y, classes, sample_weight=sample_weight))
+
+    def _check_X(self, X):
+        return check_array(X, accept_sparse=False)
+
+    def _check_X_y(self, X, y):
+        return check_X_y(X, y, accept_sparse=False)
+
+    def _check_settings(self):
+        if self.handle_unknown not in ('ignore', 'warn', 'raise'):
+            raise ValueError("The attribute 'handle_unknown' is '{}' and "
+                             "should be one of 'ignore', 'warn' or 'raise'"
+                             .format(self.handle_unknown))
+
+    def _init_counters(self, n_effective_classes, n_features):
+        self.class_count_ = np.zeros(n_effective_classes, dtype=np.float64)
+        self.category_count_ = [np.zeros((self.class_count_.shape[0], 0))
+                                for _ in range(n_features)]
+        self.feature_cat_index_mapping_ = [{} for _ in range(n_features)]
+
+    def _count(self, X, Y):
+        self.class_count_ += Y.sum(axis=0)
+        for i in range(self.n_features_):
+            X_feature = X[:, i]
+            cats = np.unique(X_feature)
+            self._update_cat_mapping(self.feature_cat_index_mapping_[i], cats)
+            # update category_count_dimensions in case partial_fit is used, to
+            # to account for unseen categories
+            self.category_count_[i] = self._update_cat_count_dims(
+                self.category_count_[i], self.feature_cat_index_mapping_[i])
+            self._update_cat_count(X_feature, Y,
+                                   self.category_count_[i],
+                                   self.feature_cat_index_mapping_[i])
+
+    def _update_cat_mapping(self, cat_mapping, cats):
```

Comment:
So you mean nested functions inside the `_count` method instead separate private methods? This would enlarge the `_count` and make it less readable in my opinion. Also, don't nested functions cause trouble when pickling them? At least I found this [stackoverflow problem](https://stackoverflow.com/questions/12019961/python-pickling-nested-functions) but in my experiments, with Python 3.7 it worked at least so maybe this information is outdated.

Reply:
pickles should be fine I think (if there's an issue the common tests should fail)

I personally find it clearer to put them inside of `_count` because it becomes obvious that these helpers are only used there and that they aren't relevant for the rest of the estimator. It helps keeping the estimator's "namespace" clean. But feel free to disagree, I won't push it.

Reply:
Alright, I also have no strong opinion about it. So we can also nest them into the `_count` method if pickling still works.

Reply:
I put the functions inside `_count`